File: lib/loader.py
import re
from .utils import note, to_lowercase, canonicalize, del_inline_comment

max_call_adr = 0x3ffff

commands = {}
datalabels = {}
disas_filename = None
result = []
labels = {}
address_requests = []
relocation_expressions = []
sizeof_cmds = []
deferred_evals = []
home = None
in_comment = False
vars_dict = {}
current_section_name = None
disasm = []

def add_command(command_dict, address, command, tags, debug_info=''):
    ''' Add a command to command_dict. '''
    assert command, f'Empty command {debug_info}'
    assert type(command_dict) is dict

    for disallowed_prefix in '0x', 'call', 'goto':
        assert not command.startswith(disallowed_prefix), \
            f'Command ends with "{disallowed_prefix}" {debug_info}'
    assert not command.endswith(':'), \
        f'Command ends with ":" {debug_info}'
    assert ';' not in command, \
        f'Command contains ";" {debug_info}'

    for prev_command, (prev_adr, prev_tags) in command_dict.items():
        if prev_command == command:
            if prev_adr == address and prev_tags == tuple(tags):
                return
            assert False, f'Command appears twice - ' \
                f'first: {prev_command} -> {prev_adr:05X} {prev_tags}, ' \
                f'second: {command} -> {address:05X} {tags} - ' \
                f'{debug_info}'

    command_dict[command] = (address, tuple(tags))

def get_commands(gadgets_file, labels_file):
    ''' Read a list of gadget names and parse a rename list. '''
    global commands, datalabels
    
    # 1. Parse gadgets file
    with open(gadgets_file, 'r', encoding='utf-8') as f:
        data = f.read().splitlines()

    in_comment = False
    line_regex = re.compile(r'([0-9a-fA-F]+)\s+(.+)')
    for line_index0, line in enumerate(data):
        line = line.strip()

        if line == '/*':
            in_comment = True
            continue
        if line == '*/':
            in_comment = False
            continue
        if in_comment:
            continue

        line = del_inline_comment(line)
        if not line:
            continue

        match = line_regex.fullmatch(line)
        if not match:
            continue
        address, command = match[1], match[2]

        command = canonicalize(command)
        command = to_lowercase(command)

        tags = []
        while command and command[0] == '{':
            i = command.find('}')
            if i < 0:
                raise Exception(f'Line {line_index0 + 1} '
                                'has unmatched "{"')
            tags.append(command[1:i])
            command = command[i + 1:]

        try:
            address = int(address, 16)
        except ValueError:
            raise Exception(f'Line {line_index0 + 1} has invalid address: {address!r}')

        add_command(commands, address, command, tags, f'at {gadgets_file}:{line_index0 + 1}')

    # 2. Parse labels file (rename list)
    with open(labels_file, 'r', encoding='u8') as f:
        data = f.read().splitlines()

    line_regex_rename = re.compile(r'^\s*([\w_.]+)\s+([\w_.]+)')
    global_regex = re.compile(r'f_([0-9a-fA-F]+)')
    local_regex  = re.compile(r'.l_([0-9a-fA-F]+)')
    data_regex   = re.compile(r'd_([0-9a-fA-F]+)')
    hexadecimal  = re.compile(r'[0-9a-fA-F]+')

    last_global_label = None
    for line_index0, line in enumerate(data):
        match = line_regex_rename.match(line)
        if not match: continue
        raw, real = match[1], match[2]
        if real.startswith('.'):
            continue
        
        match = data_regex.fullmatch(raw)
        if match:
            addr = int(match[1], 16)
            datalabels[real] = addr
            continue

        addr = None
        if hexadecimal.fullmatch(raw):
            addr = int(raw, 16)
            last_global_label = None
        else:
            match = global_regex.match(raw)
            if match:
                addr = int(match[1], 16)
                if len(match[0]) == len(raw):
                    last_global_label = addr
                else:
                    match = local_regex.fullmatch(raw[len(match[0]):])
                    if match:
                        addr += int(match[1], 16)
            else:
                match = local_regex.fullmatch(raw)
                if match:
                    if last_global_label is None:
                        print('Label cannot be read: ', line)
                        continue
                    else:
                        addr = last_global_label + int(match[1], 16)

        if addr is not None:
            assert addr < len(disasm), f'{addr:05X}'
            if disasm[addr].startswith('push lr'):
                tags = 'del lr',
                addr += 2
            else:
                tags = 'rt',
                a1 = addr + 2
                while not any(disasm[a1].startswith(x) for x in ('push lr', 'pop pc', 'rt')): a1 += 2
                if not disasm[a1].startswith('rt'):
                    tags = tags + ('del lr',)

            if real in commands:
                if 'override rename list' in commands[real][1]:
                    continue
                if commands[real] == (addr, tags):
                    note(f'Warning: Duplicated command {real}\n')
                    continue

            add_command(commands, addr, real, tags=tags,
                    debug_info=f'at {labels_file}:{line_index0+1}')
        else:
            raise ValueError('Invalid line: ' + repr(line))

def get_disassembly(filename):
    '''Try to parse a disasm file with annotated address.

    Each line should look like this:

        mov r2, 1                      ; 0A0A2 | 0201
    '''
    global disasm
    with open(filename, 'r', encoding='u8') as f:
        data = f.read().splitlines()

    # Pre-allocate array of strings to avoid resizing/appending overhead
    disasm = [''] * 262144
    for line in data:
        if line.startswith('\t') and ';' in line:
            parts = line.split(';', 1)
            instr = parts[0].strip()
            comment = parts[1]
            if '|' in comment:
                addr_hex = comment.split('|', 1)[0].strip()
                try:
                    addr = int(addr_hex, 16)
                    disasm[addr] = instr
                except ValueError:
                    pass

def sizeof_register(reg_name):
    return {'r': 1, 'e': 2, 'x': 4, 'q': 8}[reg_name[0]]

File: libcompiler_js/loader.js
import { CompilerError, note, canonicalize, del_inline_comment } from './utils.js';

let commands = {};
let datalabels = {};
let labels = {};
let vars_dict = {};
let disasm = {};
let char_to_hex = {};
let token_to_hex = {};

let home = null;
let current_section_name = null;
let in_comment = false;
let is_pass1 = false;
let current_exec_info = null;
let current_pos = 0;
let current_line_num = null;

let result = [];
let address_requests = [];
let relocation_expressions = [];
let sizeof_cmds = [];
let deferred_evals = [];
let pr_org_cmds = [];
let pr_backup_cmds = [];
let dist_cmds = [];
let backup_address = null;

let global_labels = {};
let section_addresses = {};
let label_sections = {};
let aliases = {};
let aliases_pattern = null;
let defined_functions = {};
let dynamic_macros = [];
let gadgets_offset_applied = false;

function add_command(command_dict, address, command, tags, debug_info = '') {
    if (!command || typeof command_dict !== 'object') {
        throw new CompilerError(`Empty command/dict ${debug_info}`);
    }
    if (command.startsWith('0x') || command.startsWith('call') || command.startsWith('goto')) {
        throw new CompilerError(`Command starts with disallowed ${command}`);
    }
    if (command.endsWith(':') || command.includes(';')) {
        throw new CompilerError(`Invalid command syntax ${debug_info}`);
    }
    if (command in command_dict) {
        let existing = command_dict[command];
        if (existing[0] === address && JSON.stringify(existing[1]) === JSON.stringify(tags)) {
            return;
        }
        throw new CompilerError(`Command ${command} appears twice ${debug_info}`);
    }
    command_dict[command] = [address, tags];
}

function parse_commands(gadgets_content, labels_content) {
    // We expect the file contents as strings instead of filenames
    commands = {};
    datalabels = {};

    let raw = gadgets_content.replace(/\/\*[\s\S]*?\*\//g, '');
    let gadget_lines = raw.split('\n');
    for (let i = 0; i < gadget_lines.length; i++) {
        let line = del_inline_comment(gadget_lines[i]).trim();
        if (!line) continue;

        let m = line.match(/^([0-9a-fA-F]+)\s+(.+)$/);
        if (m) {
            let addr = parseInt(m[1], 16);
            let cmd_raw = canonicalize(m[2]).toLowerCase();
            let tags = [];
            while (cmd_raw.startsWith('{')) {
                let end = cmd_raw.indexOf('}');
                if (end !== -1) {
                    tags.push(cmd_raw.substring(1, end));
                    cmd_raw = cmd_raw.substring(end + 1).trim();
                } else break;
            }
            let subs = cmd_raw.split(';').map(c => c.trim()).filter(c => c);
            for (let sub of subs) {
                add_command(commands, addr, canonicalize(sub).toLowerCase(), tags, `at gadgets line ${i + 1}`);
            }
        }
    }

    let label_lines = labels_content.split('\n');
    let last_global = null;
    for (let i = 0; i < label_lines.length; i++) {
        let line = label_lines[i];
        let m = line.match(/^\s*([\w_.]+)\s+(.+)$/);
        if (!m) continue;

        let raw_name = m[1];
        let reals = del_inline_comment(m[2]).split(';').map(r => r.trim()).filter(r => r && !r.startsWith('.'));
        if (reals.length === 0) continue;

        let d_match = raw_name.match(/^d_([0-9a-fA-F]+)$/);
        if (d_match) {
            for (let r of reals) {
                datalabels[r] = parseInt(d_match[1], 16);
            }
            continue;
        }

        let addr = null;
        if (/^[0-9a-fA-F]+$/.test(raw_name)) {
            addr = parseInt(raw_name, 16);
            last_global = null;
        } else {
            let g_match = raw_name.match(/^f_([0-9a-fA-F]+)/);
            if (g_match) {
                addr = parseInt(g_match[1], 16);
                if (g_match[0].length === raw_name.length) {
                    last_global = addr;
                } else {
                    let suffix = raw_name.substring(g_match[0].length);
                    let l_match = suffix.match(/^\.l_([0-9a-fA-F]+)$/);
                    if (l_match) {
                        addr += parseInt(l_match[1], 16);
                    }
                }
            } else {
                let l_match = raw_name.match(/^\.l_([0-9a-fA-F]+)$/);
                if (l_match && last_global !== null) {
                    addr = last_global + parseInt(l_match[1], 16);
                }
            }
        }

        if (addr !== null) {
            let tags = [];
            let dis = disasm[addr] || '';
            if (dis.startsWith('push lr')) {
                tags = ['del lr'];
                addr += 2;
            } else {
                tags = ['rt'];
                let a1 = addr + 2;
                while (a1 <= 0x3ffff) {
                    let d = disasm[a1] || '';
                    if (d.startsWith('push lr') || d.startsWith('pop pc') || d.startsWith('rt')) {
                        break;
                    }
                    a1 += 2;
                }
                if (!(disasm[a1] || '').startsWith('rt')) {
                    tags.push('del lr');
                }
            }

            for (let r of reals) {
                if (!(r in commands) || !commands[r][1].includes('override rename list')) {
                    if (r in commands && commands[r][0] === addr && JSON.stringify(commands[r][1]) === JSON.stringify(tags)) {
                        note(`Warning: Duplicated command ${r}\n`);
                        continue;
                    }
                    add_command(commands, addr, r, tags, `at labels line ${i + 1}`);
                }
            }
        }
    }
}

function parse_disassembly(disasm_content) {
    disasm = {};
    let lines = disasm_content.split('\n');
    for (let line of lines) {
        if (line.startsWith('\t') && line.includes(';')) {
            let p = line.split(';', 2);
            if (p[1].includes('|')) {
                let addrStr = p[1].split('|', 2)[0].trim();
                let addr = parseInt(addrStr, 16);
                disasm[addr] = p[0].trim();
            }
        }
    }
}

let subscript_deps = [];

function sizeof_register(reg_name) {
    const map = {
        'r': 2,
        'e': 2,
        'x': 4,
        'q': 8,
        'l': 4
    };
    return map[reg_name[0]] || 0;
}


function set_state(key, value) {
    if (key === 'home') home = value;
    if (key === 'current_section_name') current_section_name = value;
    if (key === 'in_comment') in_comment = value;
    if (key === 'is_pass1') is_pass1 = value;
    if (key === 'current_exec_info') current_exec_info = value;
    if (key === 'current_pos') current_pos = value;
    if (key === 'current_line_num') current_line_num = value;
    if (key === 'backup_address') backup_address = value;
    if (key === 'aliases_pattern') aliases_pattern = value;
    if (key === 'subscript_deps') subscript_deps = value;
}

export {
    set_state,
    commands, datalabels, labels, vars_dict, disasm, char_to_hex, token_to_hex,
    home, current_section_name, in_comment, is_pass1, current_exec_info, current_pos, current_line_num,
    result, address_requests, relocation_expressions, sizeof_cmds, deferred_evals, pr_org_cmds, pr_backup_cmds, dist_cmds, subscript_deps, backup_address,
    global_labels, section_addresses, label_sections, aliases, aliases_pattern, defined_functions, dynamic_macros, gadgets_offset_applied,
    add_command, parse_commands, parse_disassembly, sizeof_register
};
